Feature/alshulma/update stable version

alshulma opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

alshulma committed:
Update stable version

alshulma committed:
Some updates

alshulma committed:
Another change

alshulma committed:
Remove objects

alshulma committed:
Removed IncidentConfiguration

AutorestCI commented:
You don't have permission to trigger SDK Automation.
Please add yourself to ***Azure*** group from opensource portal if you are MSFT employee,
or please ask reviewer to add comment *** /openapibot sdkautomation ***.
Please ask tih@microsoft.com (or NullMDR in github) for additional help.

azuresdkci commented:
Can one of the admins verify this patch?

alshulma committed:
Fix issues

alshulma committed:
revert preview

alshulma committed:
Add objects

alshulma committed:
Add Entity

alshulma committed:
Add more

alshulma committed:
EntityCommonProperties

alshulma committed:
Changes

alshulma committed:
prettier

alshulma committed:
Remove

alshulma committed:
Make pretty

markcowl commented:
/azp run automation - sdk

ityankel committed:
Remove unneeded definitions

ityankel committed:
Update incidents definition and examples

yotam707 commented:
@hyonholee Could you assist us in this PR?

chiragg4u commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -459,6 +459,281 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.OperationalInsights/workspaces/{workspaceName}/providers/Microsoft.SecurityInsights/alertRuleTemplates": {
```
>/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.OperationalInsights/workspaces/{workspaceName}/providers/Microsoft.SecurityInsights/alertRuleTemplates": [](start = 5, length = 191)

Is this extension only applicable to OperationInsights/workspaces? Recommend adding it as a parameter.

yotam707 replied:
@ityankel

ofshomro replied:
its only on OperationInsights/workspaces

chiragg4u commented:
Please go through the contributor checklist and make sure all tools are passing before asking for review,

chiragg4u reviewed (changes requested):
Ignore this comment

ofshomro committed:
Update CreateOfficeDataConnetor.json

delete read only from data connector examples

ofshomro committed:
Update CreateOfficeDataConnetor.json

Delete read only from data connector create

ofshomro committed:
Update CreateFusionAlertRule.json

Change "true" to true

ofshomro committed:
Update CreateBookmark.json

remove type, name, id since are read only

ofshomro committed:
Update CreateActionOfAlertRule.json

remove name, id and type from example since are read only

ofshomro committed:
Update CreateBookmark.json

delete read only email and names

ofshomro committed:
Update CreateBookmark.json

remove ,

ravbhatnagar commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
       "properties": {
-        "triggerUri": {
-          "description": "Logic App Callback URL for this specific workflow.",
+        "created": {
```
is this readonly property? I am assuming this is generated by the service. ANd the user cant set this. If so, please mark readOnly:true

markcowl replied:
How was this comment resolved?

ori-licht replied:
What is the problem here?

ori-licht replied:
@markcowl , Can you please point the problem here?

markcowl replied:
This comment from the arm reviewer, that the property should be marked as read-only has not been addressed.  I assume that you resolved it some other way, and want to know what that is.

ori-licht replied:
Hi,
the triggerUri is not a readonly property. the user supplied this property and the user can change it as well

ravbhatnagar commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
+          "description": "The query result of the bookmark.",
+          "type": "string"
+        },
+        "updated": {
```
readonly?

markcowl replied:
same comment - if this is server generated, it must be marked as readonly

ravbhatnagar reviewed (commented):
A couple of comments. Rest looks ok. Signing off from ARM side.

chiragg4u reviewed (approved):
Approving based on Gaurav's feedback,.

markcowl commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
+          "description": "Etag of the action.",
+          "type": "string"
+        },
+        "properties": {
```
why are there 2 definitions for 'properties'?

markcowl commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
-          "description": "AATP (Azure Advanced Threat Protection) data connector properties.",
-          "x-ms-client-flatten": true
+        "alerts": {
+          "allOf": [
```
Why use an allof for a property definition? Use a reference.

markcowl commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
+      "description": "The available data types for Amazon Web Services CloudTrail data connector.",
+      "properties": {
+        "logs": {
+          "allOf": [
```
Use a reference, not an allof

markcowl commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
-        }
-      ],
-      "description": "Action property bag.",
+    "CloudError": {
```
Please use the common ddefintion, like this: https://github.com/Azure/azure-rest-api-specs/blob/9df7c71f9b955ea66fc9ede43842f937ca53a646/specification/resources/resource-manager/Microsoft.Resources/stable/2019-08-01/resources.json#L4232

laithhisham replied:
done

laithhisham replied:
Apparently, this is a breaking change - the breaking change check is failing after that change
is it fine @markcowl ?

markcowl commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -1030,378 +2045,529 @@
+        }
+      },
+      "type": "object",
+      "x-ms-discriminator-value": "Fusion"
```
This property must be required

markcowl replied:
the property that contaisn the discriminator must be required.  That is, if I have a property called 'type'  whose value is the discriminator value, the 'type' property must be required.

laithhisham replied:
not sure I get it - "Fusion" is not a property
If I am not mistaken the x-ms-discriminator-value defines an alias of the type
am I missing something?

markcowl replied:
In this case, it means the discriminator from the parent type, which is, in fact required.  So, you can disregard

markcowl replied:
Yes, fudion is the value, 'kind' is the property, and it looks like it is requred in the referenced type

markcowl commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -1030,378 +2045,529 @@
+      "description": "Represents Fusion alert rule template.",
       "properties": {
-        "alerts": {
+        "properties": {
```
do not use an allof for properties, use a ref to the defining object

laithhisham replied:
done

laithhisham replied:
Apparently, this is a breaking change - the breaking change check is failing after that change
is it fine @markcowl ?

markcowl replied:
That's unfortunate.  It should be equivalent, but it's OK to switch back to avoid the breaking change

laithhisham replied:
I am afraid we shouldn't switch back
There is a problem in the previous definition with the nested properties definition, which was not correct. Switching back to the allOf in the property forces us to return the nested properties field.

Shall we keep it like that.. I think it is not the only Breaking Change anyway.. The CloudError that you suggested to change (to use the common one) is also a breaking change

markcowl replied:
We should make sure the swagger spec is correct, so if the previous definition was incorrect, we should correct it.

On CloudError, this is a usability problem with the errors, but if it would actually be a breaking change, then we should not make the breaking change.

markcowl commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -1030,378 +2045,529 @@
-          "description": "Alerts data type connection.",
-          "type": "object"
+          "description": "Fusion alert rule template properties",
+          "properties": {
```
cannot have 2 definitions of properties

laithhisham replied:
done

ityankel committed:
Update operationId of GET on a single incident comment

Co-authored-by: Mark Cowlishaw <markcowl@microsoft.com>

ityankel commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
+          "IncidentComments"
+        ],
+        "description": "Gets all incident comments.",
+        "operationId": "IncidentComments_ListByIncident",
```
We called this operation 'list by incident', because we're still not sure if we'll also add operations to list comments by other scopes (e.g list all comments under the workspace) and we wanted the name for this operation to be indicative so we won't need to change it if/when we add more list operations.
Do you think it's better to call it 'IncidentComments_List' for now and rename it later if we add more list operations?

ofshomro committed:
Update SecurityInsights.json

Add "description": "OK, Operation successfully completed",

ofshomro commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -459,6 +459,281 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
```
done

ofshomro commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -459,6 +459,281 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
```
done

ofshomro commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -459,6 +459,281 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
```
done

ofshomro commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -459,6 +459,281 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
```
done

ofshomro commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
```
done

ofshomro committed:
Update SecurityInsights.json

Remove redundant ,

laithhisham committed:
Fixed Analytics issues + CloudError from common

laithhisham committed:
More fixes to the Analytics sections

laithhisham committed:
fixed AlertRuleKind after referencing to it - changed to string

laithhisham commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
+      "type": "object"
+    },
+    "AlertRuleTemplate": {
+      "allOf": [
```
done

laithhisham committed:
fix incorrect hierarchy in MicrosoftSecurityIncidentCreationAlertRuleTemplate

laithhisham committed:
extracted MicrosoftSecurityIncidentCreationAlertRuleTemplateCommonProperties to a separate object definition

laithhisham committed:
extracted MicrosoftSecurityProductName to enum

yotam707 commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
+    },
+    "AADDataConnectorProperties": {
+      "allOf": [
+        {
```
@markcowl, we have multiple entries that has more than a single resource of `allOf` .What is the logic behind this? Why it this an array if it can only have a single property?

markcowl replied:
There are many things that are allowed in Swagger, which are not supported by all languages.  In this case, the issue is that, for languages that don't support multiple inheritance, multiple AllOf cannot be modeled

yotam707 replied:
I see that a lot of other API versions have multiple `allOf`. This is the API and we need the TenantId and the alert properties. So why this can't be modeled? I don't understand.

ofshomro committed:
Update SecurityInsights.json

"discriminator": "kind", in AlertRuleKind and DataConnectorKind

markcowl commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
-      "type": "object"
+      "type": "string",
+      "x-ms-enum": {
+        "modelAsString": false,
```
Changing a value in an enum is a breakigng change.  This is why most 'enums' use x-ms-enum with 'modelAsString' set to true - this is modeled as an extensible enum, where the customer is not limited to an 'exhaustive' set of values, and therefore adding values is *not* a breaking change.

laithhisham replied:
@markcowl - so you suggest changing the "modelAsString" to true?

ofshomro committed:
Update SecurityInsights.json

Kind discriminator required

laithhisham commented on specification/securityinsights/resource-manager/Microsoft.SecurityInsights/stable/2020-01-01/SecurityInsights.json:
```diff
@@ -664,364 +939,1104 @@
+    },
+    "AlertRule": {
+      "allOf": [
+        {
```
done

laithhisham committed:
remove discriminator and required from enum - it is set on the using definitions

laithhisham committed:
Revert "remove discriminator and required from enum - it is set on the using definitions"

This reverts commit c24e7d0a85c9cbc5025c2b4f8cc80800df66c44c.

Revert "extracted MicrosoftSecurityProductName to enum"

This reverts commit 2389fc968e477d5ef300f5b71cc61a8d23bcb5f1.

Revert "extracted MicrosoftSecurityIncidentCreationAlertRuleTemplateCommonProperties to a separate object definition"

This reverts commit 72280091f402ba9e7cb2be3f62cd16cb58b99188.

Revert "fix incorrect hierarchy in MicrosoftSecurityIncidentCreationAlertRuleTemplate"

This reverts commit dbf12a9189c3718f49229fe8a6a5ffa99150615c.

revert 03ace77f5

revert 1cc27d544

Revert "Fixed Analytics issues + CloudError from common"

This reverts commit 39bffde1aaebe654d2c0b09dc9c9e84b1f9fe911.

Revert Analytics Changes

laithhisham committed:
revert the discriminator kind change

laithhisham committed:
Revert "revert the discriminator kind change"

This reverts commit abee5fb24694045780861ac978ac87fe4ca88eeb.

laithhisham committed:
Revert "Revert "remove discriminator and required from enum - it is set on the using definitions""

This reverts commit 81054a97bdf1417b533f2c989563b8a826520361.

ofshomro committed:
Update SecurityInsights.json

add "format": "int32" to integer types

Commit:
DataConnectorKind to enum

Change DataConnectorKind to enum and fix DataConnector to use it in the property

Commit:
SettingsKind to enum

Change SettingsKind to enum and use in the Settings properties.

Commit:
Remove double allOff in connectors

Commit:
Fix connectors inner properties

laithhisham committed:
fix scheduled alert rule template properties

laithhisham committed:
fixed nested all of issue in analytic rules templates

Commit:
Fix inner allof in connectors properties

Commit:
Merge branch 'feature/alshulma/UpdateStableVersion' of https://github.com/alshulma/azure-rest-api-specs into feature/alshulma/UpdateStableVersion
